shared-bindings/_protomatter/Protomatter.c

```diff
@@ -0,0 +1,362 @@
+}
+
+STATIC void validate_pins(qstr what, uint8_t* pin_nos, mp_int_t max_pins, mp_obj_t seq, uint8_t *count_out) {
+    mp_int_t len = MP_OBJ_SMALL_INT_VALUE(mp_obj_len(seq));
+    if (len > max_pins) {
+        mp_raise_ValueError_varg(translate("At most %d %q may be specified (not %d)"), max_pins, what, len);
+    }
+    *count_out = len;
+    for (mp_int_t i=0; i<len; i++) {
+        pin_nos[i] = validate_pin(mp_obj_subscr(seq, MP_OBJ_NEW_SMALL_INT(i), MP_OBJ_SENTINEL));
+    }
+}
+
+STATIC void claim_and_never_reset_pin(mp_obj_t pin) {
+    common_hal_mcu_pin_claim(pin);
+    common_hal_never_reset_pin(pin);
+}
+
+STATIC void claim_and_never_reset_pins(mp_obj_t seq) {
+    mp_int_t len = MP_OBJ_SMALL_INT_VALUE(mp_obj_len(seq));
+    for (mp_int_t i=0; i<len; i++) {
+        claim_and_never_reset_pin(mp_obj_subscr(seq, MP_OBJ_NEW_SMALL_INT(i), MP_OBJ_SENTINEL));
+    }
+}
+
+STATIC void preflight_pins_or_throw(uint8_t clock_pin, uint8_t *rgb_pins, uint8_t rgb_pin_count, bool allow_inefficient) {
+    uint32_t port = clock_pin / 32;
+    uint32_t bit_mask = 1 << (clock_pin % 32);
+
+    for (uint8_t i = 0; i < rgb_pin_count; i++) {
+        uint32_t pin_port = rgb_pins[i] / 32;
+
+        if (pin_port != port) {
+            mp_raise_ValueError_varg(
+                translate("rgb_pins[%d] is not on the same port as clock"), i);
+        }
+
+        uint32_t pin_mask = 1 << (rgb_pins[i] % 32);
+        if (pin_mask & bit_mask) {
+            mp_raise_ValueError_varg(
+                translate("rgb_pins[%d] duplicates another pin assignment"), i);
+        }
+
+        bit_mask |= pin_mask;
+    }
+
+    if (allow_inefficient) {
+        return;
+    }
+
+    uint8_t byte_mask = 0;
+    if (bit_mask & 0x000000FF) byte_mask |= 0b0001;
+    if (bit_mask & 0x0000FF00) byte_mask |= 0b0010;
+    if (bit_mask & 0x00FF0000) byte_mask |= 0b0100;
+    if (bit_mask & 0xFF000000) byte_mask |= 0b1000;
+
+    uint8_t bytes_per_element = 0xff;
+    uint8_t ideal_bytes_per_element = (rgb_pin_count + 7) / 8;
+
+    switch(byte_mask) {
+        case 0b0001:
+        case 0b0010:
+        case 0b0100:
+        case 0b1000:
+            bytes_per_element = 1;
+            break;
+
+        case 0b0011:
+        case 0b1100:
+            bytes_per_element = 2;
+            break;
+
+        default:
+            bytes_per_element = 4;
+            break;
+    }
+
+    if (bytes_per_element != ideal_bytes_per_element) {
+        mp_raise_ValueError_varg(
+            translate("Pinout uses %d bytes per element, which consumes more than the ideal %d bytes.  If this cannot be avoided, pass allow_inefficient=True to the constructor"),
+            bytes_per_element, ideal_bytes_per_element);
+    }
+}
+
+//| :class:`~_protomatter.Protomatter` displays an in-memory framebuffer to an LED matrix.
+//|
+//| .. class:: Protomatter(width, bit_depth, rgb_pins, addr_pins, clock_pin, latch_pin, output_enable_pin, *, doublebuffer=True, framebuffer=None)
```

Comment:
I was just peeking at your guide for how I can test this. I think we want these all to be kwarg-only args because they are either numbers, pins or lists of pins. Reading the code doesn't help with what is what. The current positional arguments can be required and have no default.

I'd also take in the height even though it's implicit in the address pins. It can make the framebuffer smaller than the available address pins and can also validate enough address pins are given.

Reply:
Yeah, I can make them kw-only.  

I can make it so that height, if specified, is checked against the expected number.  I'm not sure of the use case of creating a 16 line display on a 32 line panel, though.

*however*, the raspberry pi lib for driving these displays (I think that's the one) can drive multiple displays wired in "zigzag" fashion.  In that case, the relation between the WxH of the screen is decoupled from the true width of the display.  Protomatter C library doesn't support this (yet, not sure if planned).